docs/ldml/tr35-keyboards.md

```diff
@@ -308,6 +308,77 @@ For extensibility, the `<special>` element will be allowed at nearly every level
 
 See [Element special](tr35.md#special) in Part 1.
 
+## Normalization
+
+Normalization will not typically be the responsibility of the keyboard author, rather this will be managed by the implementation.
+The implementation will apply normalization as appropriate when matching transform rules and `<display>` value matching.
+Output from the keyboard, following application of all transform rules, will be normalized to implementation- or application-requested form.
+
+The attribute value `normalization="disabled"` can be used to indicate that no automatic normalization happens.
+Using this setting is not recommended. See [`<settings>`](#element-settings) for details.
+
+Input source files may be in any normalization format, however, authors should be aware of two areas where normalization affects keyboard operation: that of transform matching, and that of output.
+
+Normalization can be applied within segments defined by Grapheme Cluster Boundaries as defined by [UAX #29](https://www.unicode.org/reports/tr29/#Grapheme_Cluster_Boundaries). This segmentation is pertinent where markers are involved, see below.
+
+### Normalization and Transform Matching
+
+Regardless of the normalization form in the keyboard source file or in the edit buffer context, transform matching will be performed using **NFD**. For example, all of the following transforms will match the input strings `è̠`, whether the input is U+00E8 U+0320, U+0065 U+0320 U+0300, or U+0065 U+0300 U+0320.
+
+```xml
+<transform from="e\u{0300}\u{0320}" /> <!-- Non-NFD but will match -->
+<transform from="e\u{0320}\u{0300}" />
+<transform from="\u{00E8}\u{0320}" /> <!-- è + U+0320 -->
+```
+
+#### Normalization and Markers
+
+A special issue occurs when markers are involved. The markers may be reordered along with characters.
+
+**Example 1**
+
+Consider this example, without markers:
+
+- `e\u{0300}\u{0320}` (original)
+- `e\u{0320}\u{0300}` (NFD)
+
+If we add markers:
+
+- `e\u{0300}\m{marker}\u{0320}` (original)
+- `e\m{marker}\u{0320}\u{0300}` (NFD)
+
+The principle is that the marker stays 'glued' to the following character, in this case the `\u{0320}`. If a marker occurs at the end of input or at the end of a segment (a grapheme cluster boundary), the marker is 'glued' to the end of that segment during the course of normalization. As another example:
```

Comment:
> [...] or at the end of a segment (a grapheme cluster boundary), [...]

Does this mean that a marker can never be directly associated with what becomes the first character of a segment / grapheme cluster?

What if the following happens?  

User types (`\m{marker}` + `k`) to start a new cluster.
- with the marker and `k` output from a single keystroke

Does that `\m{marker}` now associate with the end of the prior segment?

----

Furthermore, suppose a language where vowels cluster with consonants, the latter of which should be first in the cluster.

User types: `e` (`\m{marker_1}` + `k`)
- with the marker and `k` output from a single keystroke
Normalization reorder: `\m{marker_1} k e`

Does _that_ `\m{marker}` now associate with the end of the prior segment?

----

Finally...

User types: (`\m{marker_2}` `e`) (`\m{marker_1}` + `k`)
- with the marker and `k` output from a single keystroke
Normalization reorder (if single-segment only): `\m{marker_1} k \m{marker_2} e`
- but... that `\m{marker_2}` was temporarily at the segment boundary before it got repositioned.

My interpretation of the spec as stated above, if a prior segment exists, would suggest...

(first keystroke)
`<prior segment>` `\m{marker_2}` `e`
- `{\m{marker_2}` becomes bound to prior segment

(second keystroke)
`<prior segment>` `\m{marker_2}` \m{marker_1} `k` `e`
- with `\m{marker_1}` additionally bound to the prior segment

----

Of course, my interpretation here could be wrong. 

Also, the logic that would be required to ensure that such an `e` (attaching, not base) doesn't attach to the prior segment before the `k` appears would likely be interesting, and I've kind of handwaved that here for the sake of this example.

Reply:
> > [...] or at the end of a segment (a grapheme cluster boundary), [...]
> 
> Does this mean that a marker can never be directly associated with what becomes the first character of a segment / grapheme cluster?

The 'association' here is just within a single step of processing. If something else is typed, then the text is re-analyzed.  The 'glue' does not persist.

> What if the following happens?
> 
> User types (`\m{marker}` + `k`) to start a new cluster.
> 
> * with the marker and `k` output from a single keystroke
> * `k` as a base character, validly starting a new cluster re: normalization.
> 
> Does that `\m{marker}` now associate with the end of the prior segment?

I think the answer is yes here.  Since `k` is a new cluster, the marker is at the 'end of the segment' and so will remain at the end, that is, that it won't be rearranged. In other words, the marker is in the same place regardless of prior chars, and regardless of whether the `k` was there or not. 

> Furthermore, suppose a language where vowels cluster with consonants, the latter of which should be first in the cluster.

Now I need to interrupt here, though- cluster *does not refer to any sort of tailored clusters*. It is not in any way language dependent.  So the following would never happen with `k` or `e`.   

"grapheme cluster" may be incorrect here, and should refer to normalization boundaries specifically.

> User types: `e` (`\m{marker_1}` + `k`)
> 
> * with the marker and `k` output from a single keystroke
> * `e` as an 'attaching' glyph, not a base glyph - same grapheme cluster
>   Normalization reorder: `\m{marker_1} k e`
> 
> Does _that_ `\m{marker_1}` now associate with the end of the prior segment?

If `e` and `k` were instead combining marks for example, it might reorder this way.  If "`k`" moves, then marker_1 moves.

> Finally...
> 
> User types: (`\m{marker_2}` `e`) (`\m{marker_1}` + `k`)
> 
> * with the marker and `k` output from a single keystroke
> * and marker + `e` output from a single keystroke
>   Normalization reorder (if single-segment only): `\m{marker_1} k \m{marker_2} e`
> * but... that `\m{marker_2}` was temporarily at the segment boundary before it got repositioned.

Segment boundaries don't persist.  If `k` would reorder before `e` (which it won't for NFD), then the markers would move as you noted.  

Also, the keystroke order isn't significant here.  Normalization is applied after the keystroke appends to the input text.  Whether one or 5 codepoints were added does not affect the normalization processing.  

> 
> My interpretation of the spec as stated above, if a prior segment exists, would suggest...
> 
> (first keystroke) `<prior segment>` `\m{marker_2}` `e`
> 
> * `{\m{marker_2}` becomes bound to prior segment
> 
> (second keystroke) `<prior segment>` `\m{marker_2}` `\m{marker_1}` `k` `e`
> 
> * with `\m{marker_1}` additionally bound to the prior segment
> 
> Of course, my interpretation here could be wrong.
> 
> Also, the logic that would be required to ensure that such an `e` (attaching, not base) doesn't attach to the prior segment before the `k` appears would likely be interesting, and I've kind of handwaved that here for the sake of this example.

I don't think this is needed since the boundary analysis is done after text is added and doesn't persist.

Reply:
```suggestion
The principle is that the marker stays 'glued' to the following character, in this case the `\u{0320}`. If a marker occurs at the end of input or at the end of a segment (a grapheme cluster boundary), the marker is 'glued' to the end of that segment during the course of normalization for that keystroke. As another example:
```

As a normalization pass triggered later, by a followup keystroke, can reanchor the marker, it might be good to imply this occurs.  I certainly didn't infer it when I wrote the first comment in this chain.

Not sure if this is the _best_ wording, of course - feel free to workshop that.

Reply:
temporarily glued? "for the current keystroke" is actually not quite correct, because multiple normalizations may occur during keystroke transform processing, and the gluing has to be done for each round of normalization; it's never stored.

Reply:
i think it's an improvement,thanks..